### PR TITLE
Fixing the implementation of supertraits to match Rust

### PR DESCRIPTION
**Pull request by mohammadfawaz**
My original implementation of traits did not align with Rust. I previously expected the trait and its supertraits to all be implemented in a single `impl` block. Rust expects that each trait continues to be implemented in its own `impl` block. If a trait `B` has a supertrait `A`, this really means that if `B` is implemented for a type `T`, then `A` has to also be implemented for `T`.

Two missing things:
* There are still some issues around having multiple functions with the same name defined for the same type. That may be through a trait/supertrait relationship or even the same function defined twice for a single trait. I'm openning an issue about that. I think in some cases, this can be resolved with having call paths directly in the code (e.g. https://github.com/FuelLabs/sway/pull/472). Issue opened: https://github.com/FuelLabs/sway/issues/837
* This case:
```rust
trait A {
    fn f(self) -> u64;
} {
    fn add_f(self, x: u64) -> u64 {
        self.f() + x
    }
}

trait B: A {
    fn g(self) -> u64;
} {
    fn add_g(self, x: u64) -> u64 {
       // Should this function have access to `f()` from `trait A`?        
    }
}
```

Also closes https://github.com/FuelLabs/sway/issues/767

### CODE REVIEW

**Comment by adlerjohn** [comment]
Should we merge from `master` to pick up changes from #472?

**Comment by mohammadfawaz** [comment]
> Should we merge from `master` to pick up changes from #472?

Done, [#472](https://github.com/FuelLabs/sway/pull/472) gets us closer but is not enough yet. I summarized everything here: https://github.com/FuelLabs/sway/issues/837

File: sway-core/src/semantic_analysis/syntax_tree.rs [hunk 1]
```diff
@@ -262,6 +267,82 @@ impl TypedParseTree {
...
+                            }
+                        }
+                        false
+                    });
+                    if impl_of_supertrait.is_none() {
+                        // The two errors below should really be a single error (and a "note"),
+                        // but we don't have a way today to point to two separate locations in the
+                        // user code with a single error.
+                        errors.push(CompileError::SupertraitImplMissing {
+                            supertrait_name: supertrait.name.span().span.as_str().to_string(),
```
Comment (otrho) [inline sway-core/src/semantic_analysis/syntax_tree.rs:329]: This is a bit crazy!  Is this the same?
```suggestion
                            supertrait_name: supertrait.name.to_string(),
```
`CallPath` implements `Display` and I think includes the prefix + suffix.

**Review by otrho — CHANGES_REQUESTED** [review]

**Comment by mohammadfawaz** [comment]
@otrho, I addressed all your comments. Thanks for the suggestions!

File: sway-core/src/semantic_analysis/syntax_tree.rs (hunk 1 above)
Comment (mohammadfawaz) [reply sway-core/src/semantic_analysis/syntax_tree.rs]: Indeed! Your suggestions is a lot better.

**Review by sezna — APPROVED** [review]

**Review by otrho — APPROVED** [review]

**Pull request merged**